core/storage/btree.rs

```diff
@@ -1755,138 +1755,169 @@ impl BTreeCursor {
+                        )?;
 
-            // 2. Copy that entry to replace the one being deleted
-            let leaf_contents = leaf_page.get().contents.as_ref().unwrap();
-            let leaf_cell_idx = self.stack.current_cell_index() as usize - 1;
-            let predecessor_cell = leaf_contents.cell_get(
-                leaf_cell_idx,
-                self.pager.clone(),
-                payload_overflow_threshold_max(
-                    leaf_contents.page_type(),
-                    self.usable_space() as u16,
-                ),
-                payload_overflow_threshold_min(
-                    leaf_contents.page_type(),
-                    self.usable_space() as u16,
-                ),
-                self.usable_space(),
-            )?;
+                        // 3. Create an interior cell from the leaf cell
+                        let mut cell_payload: Vec<u8> = Vec::new();
+                        match predecessor_cell {
+                            BTreeCell::TableLeafCell(leaf_cell) => {
+                                // Format: [left child page (4 bytes)][rowid varint]
+                                if let Some(child_pointer) = original_child_pointer {
+                                    cell_payload.extend_from_slice(&child_pointer.to_be_bytes());
+                                    write_varint_to_vec(leaf_cell._rowid, &mut cell_payload);
+                                }
+                            }
+                            _ => unreachable!("Expected table leaf cell"),
+                        }
+                        insert_into_cell(
+                            contents,
+                            &cell_payload,
+                            cell_idx,
+                            self.usable_space() as u16,
+                        )
+                        .unwrap();
+                        drop_cell(contents, cell_idx, self.usable_space() as u16)?;
+                    } else {
+                        // For leaf nodes, simply remove the cell
+                        drop_cell(contents, cell_idx, self.usable_space() as u16)?;
+                    }
+
+                    let page = self.stack.top();
+                    return_if_locked!(page);
 
-            // 3. Create an interior cell from the leaf cell
-            let mut cell_payload: Vec<u8> = Vec::new();
-            match predecessor_cell {
-                BTreeCell::TableLeafCell(leaf_cell) => {
-                    // Format: [left child page (4 bytes)][rowid varint]
-                    if let Some(child_pointer) = original_child_pointer {
-                        cell_payload.extend_from_slice(&child_pointer.to_be_bytes());
-                        write_varint_to_vec(leaf_cell._rowid, &mut cell_payload);
+                    if !page.is_loaded() {
+                        self.pager.load_page(page.clone())?;
+                        return Ok(CursorResult::IO);
+                    }
```

Comment:
Yeah we have another one of these, I imagine we need to have more Delete states to really make this re-entrant.